### PR TITLE
1st Deliverable - Go Workshop - Oscar Santoyo - mentor: Roboam

### PR DESCRIPTION
**Pull request by oscarSantoyo**
This PR contains requirements: 

- [X] Create API
- [X] Add an endpoint to read from a CSV file
- [X] The CSV should have information including ID
- [X] Endpoint to get information from CSV filtered by some field
- [X] The result of filtering is displayed as response
- [ ] Clean architecture proposal - I didn't use the resources from classroom since I believe is a really overwork that can be accomplish using an IoC container, so I used an IoC container instead - let's discuss if needed
- [X] Best practices
- [X] Handle errors

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,23 @@
...
+
+### Short architecture explanation
+
+Using `container` as IoC provider to rely on interfaces not hard structs implementations
+
+- main.go :: just creates the server, wires IoC container
+- router.go :: sets the routes and wires controllers to it
+- controller :: controllers here, wires endpoint to services using `container` IoC container 
+- service :: atomic functionality to solve one necessity.
+
```
Comment (RolloCasanova) [inline README.md:35]: Please, add some relevant information regarding your project, what it does, and endpoints it handles.

File: container/inject.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package container
+
+import (
+	"fmt"
+
+	"github.com/golobby/container"
+	"github.com/oscarSantoyo/academy-go-q12021/service"
+)
```
Comment (RolloCasanova) [inline container/inject.go]: As a best practice, we should organize our imports in alphabetic order and taking first into consideration the following groups separated by an empty line:

First, Go's standard libraries (fmt)
Second, your local packages (github.com/oscarSantoyo/academy-go-q12021/service)
And lastly, 3rd party packages (github.com/golobby/container)

Something like this:
```suggestion
import (
	"fmt"

	"github.com/oscarSantoyo/academy-go-q12021/service"
	
	"github.com/golobby/container"
)
```

File: container/inject.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+func Connect ( ){
+	fmt.Println("Connecting container")
```
Comment (RolloCasanova) [inline container/inject.go]: Although this is okay to have these prints on the code, this should be done through a logger. 

Please, give it a look to `log`, `logrus` or `zerolog` packages for this.

File: container/inject.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+	// var instance = container.NewContainer();
```
Comment (RolloCasanova) [inline container/inject.go]: Normally, comments are okay to be in code to add a description, or to have some explanation about something inside the code that might not be clear enough. Please, remove this commented line, as some linters may mark this as an error.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+func startServer() {
+	e := echo.New()
+
+	e.Use(middleware.Logger())
+	e.Use(middleware.Recover())
+	router.SetRouter(e)
+
+	e.Logger.Fatal(e.Start(":8080"))
```
Comment (RolloCasanova) [inline main.go]: According to clean architecture, this port value should be read from a config file. Please, give it a look to https://github.com/spf13/viper to know about configuration files in go.

File: main_test.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+)
+
+func TestUnmarshallSearchResponse(t *testing.T) {
+	var searchData SearchResponse
```
Comment (RolloCasanova) [inline main_test.go:9]: SearchResponse doesn't exist in the main package; I think you wanted to use it from service, so first import your service package, and then make a reference to SearchResponse from service as `service.SearchResponse`.

File: main_test.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+	testData := "{\"numFound\": 100,\"start\": 0,\"docs\": [{ \"cover_i\": 855045,\"has_fulltext\": false,\"title\": \"Untangling Tolkien\",\"title_suggest\": \"Untangling Tolkien\",           \"type\": \"work\",            \"ebook_count_i\": 0,            \"edition_count\": 1,            \"key\": \"/works/OL5747274W\",\r\n            \"last_modified_i\": 1383146561,\r\n            \"cover_edition_key\": \"OL3319316M\",\r\n            \"first_publish_year\": 2003\r\n}]}"
+	json.Unmarshal([]byte(testData), &searchData)
+	fmt.Println(searchData.Start)
+	fmt.Println(searchData.NumFound)
+	fmt.Println(len(searchData.Doc))
+	// fmt.Println(searchData.Doc[0].Key)
```
Comment (RolloCasanova) [inline main_test.go]: Please, remove this comment.

File: main_test.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+	"testing"
+)
+
+func TestUnmarshallSearchResponse(t *testing.T) {
+	var searchData SearchResponse
+	testData := "{\"numFound\": 100,\"start\": 0,\"docs\": [{ \"cover_i\": 855045,\"has_fulltext\": false,\"title\": \"Untangling Tolkien\",\"title_suggest\": \"Untangling Tolkien\",           \"type\": \"work\",            \"ebook_count_i\": 0,            \"edition_count\": 1,            \"key\": \"/works/OL5747274W\",\r\n            \"last_modified_i\": 1383146561,\r\n            \"cover_edition_key\": \"OL3319316M\",\r\n            \"first_publish_year\": 2003\r\n}]}"
+	json.Unmarshal([]byte(testData), &searchData)
+	fmt.Println(searchData.Start)
+	fmt.Println(searchData.NumFound)
+	fmt.Println(len(searchData.Doc))
```
Comment (RolloCasanova) [inline main_test.go]: Not sure about these lines, but technically speaking, there's nothing being tested here. 

Give it a look to `testing` and `github.com/stretchr/testify/assert` packages to create some assertions and proper testing.

P.D. If using VS Code, there's a trick to create a test template (for function, file, or package). Cmd + Shift + P, then write `Generate unit tests for` and select proper action.

File: service/csvService.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+// 	if _, err := os.Stat(file); os.IsNotExist(err) {
+// 		fmt.Println("File "+file+" does not exists")
+// 		return false
+// 	}
+// 	return true
+// }
+
+// func readFileByRow(file string) ([]string, error) {
+
+// }
```
Comment (RolloCasanova) [inline service/csvService.go]: Remove this commented lines

File: service/csvService.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+	var responseObject SearchResponse
+
+	json.Unmarshal(responseData, &responseObject)
+
+	file, err := os.Create(CSVFile)
+	defer file.Close()
+
+	if err != nil {
+		log.Fatal("Cannot create file", err)
```
Comment (RolloCasanova) [inline service/csvService.go:96]: If this happens, defer on line 111 will not run.

File: service/csvService.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+}
+
+func SaveCsv(responseObject SearchResponse, file os.File) {
+	writer := csv.NewWriter(&file)
+	defer writer.Flush()
+	for _, doc := range responseObject.Doc {
+		err := writer.Write(interfaceToString(structs.Values(doc)))
+		if err != nil {
+			log.Fatal("cannot write CSV", err)
```
Comment (RolloCasanova) [inline service/csvService.go:109]: So does here, if this happends, defer on line 122 will not run.

File: service/csvService.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+type Doc struct {
+	Key       string `json:"key"`
+	Title     string `json:"title"`
+	Type      string `json:"type"`
+	Published string `json:"first_published_year"`
+}
+
+var CSVFile = "/home/omar/Desktop/saved.csv"
+
+// func (c )LoadCsvData
```
Comment (RolloCasanova) [inline service/csvService.go]: As a best practice, all functions should contain a comment in the function to describe the function's purpose.

This should be done by adding a space, then the name of the function, and after that a brief explanation.

As an example:
```
// areEqual returns if x and y are the same value, returning an error if ...
func areEqual(x, y int) (bool, error) {
... 
}
```

Something like that.

File: service/csvService.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+func (c CsvServiceImpl) DownloadCsvData() error {
+	getAndSaveData()
+	return nil
+}
+
+func (c CsvServiceImpl) FilterById(id string) ([]Doc, error) {
```
Comment (RolloCasanova) [inline service/csvService.go]:
```suggestion
func (c CsvServiceImpl) FilterByID(id string) ([]Doc, error) {
```

File: service/csvService.go [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+type Doc struct {
+	Key       string `json:"key"`
+	Title     string `json:"title"`
+	Type      string `json:"type"`
+	Published string `json:"first_published_year"`
+}
+
+var CSVFile = "/home/omar/Desktop/saved.csv"
```
Comment (RolloCasanova) [inline service/csvService.go]: This should not be hardcoded. Instead, use a configuration file to read this value.

Add this as a dependency of your service struct as a struct's field.

Also, I think it's a better idea to have a relative route to the project (and include the file) instead to an absolute route.

File: service/csvService.go [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+	NumFound int   `json:"numFound"`
+	Start    int   `json:"start"`
+}
+
+type Doc struct {
+	Key       string `json:"key"`
+	Title     string `json:"title"`
+	Type      string `json:"type"`
+	Published string `json:"first_published_year"`
+}
```
Comment (RolloCasanova) [inline service/csvService.go]: According to clean architecture, this should be moved into a `model` package.

File: service/csvService.go [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+	for _, doc := range responseObject.Doc {
+		err := writer.Write(interfaceToString(structs.Values(doc)))
+		if err != nil {
+			log.Fatal("cannot write CSV", err)
+		}
+	}
+}
+
+func interfaceToString(record []interface{}) []string {
+	var a []string
```
Comment (RolloCasanova) [inline service/csvService.go]: You could initialize it, as you already know how many records you have. Even better, no need to specify the type of a.

```suggestion
	a  := make([]string, len(record))
```

File: service/searchService.go [hunk 17]
```diff
@@ -0,0 +1,29 @@
+package service
+
+import (
+	"fmt"
+	"github.com/golobby/container"
+)
```
Comment (RolloCasanova) [inline service/searchService.go]:
```suggestion
import (
	"fmt"
	
	"github.com/golobby/container"
)
```

File: service/csvService.go [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+// 		return false
+// 	}
+// 	return true
+// }
+
+// func readFileByRow(file string) ([]string, error) {
+
+// }
+func getAndSaveData() {
+
```
Comment (RolloCasanova) [inline service/csvService.go]: Unnecessary leading empty line
```suggestion
```

File: service/csvService.go [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+	json.Unmarshal(responseData, &responseObject)
+
+	file, err := os.Create(CSVFile)
+	defer file.Close()
+
+	if err != nil {
+		log.Fatal("Cannot create file", err)
+	}
+	SaveCsv(responseObject, *file)
+
```
Comment (RolloCasanova) [inline service/csvService.go:100]:
```suggestion
```

File: service/csvService.go [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+// func (c )LoadCsvData
+func (c CsvServiceImpl) DownloadCsvData() error {
+	getAndSaveData()
+	return nil
+}
+
+func (c CsvServiceImpl) FilterById(id string) ([]Doc, error) {
+	file, err := os.Open(CSVFile)
+	if err != nil {
+		return nil, err
```
Comment (RolloCasanova) [inline service/csvService.go]: When working with external packages, it's better to wrap the error to add a specific error message in this layer.

```suggestion
		return nil, fmt.Errorf("failed to open file: %v", err)
```

File: service/csvService.go [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+	reader := csv.NewReader(file)
+	records, errReader := reader.ReadAll()
+	if errReader != nil {
+		return nil, errReader
+	}
+
+	var recordsStruct []Doc
+
+	for _, record := range records {
+		fmt.Println(record)
```
Comment (RolloCasanova) [inline service/csvService.go]: log this instead of using print.

File: service/csvService.go [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+
+		doc := Doc{
+			Key:       record[0],
+			Title:     record[1],
+			Type:      record[2],
+			Published: record[3],
+		}
+		if doc.Key == id {
+			recordsStruct = append(recordsStruct, doc)
+		}
```
Comment (RolloCasanova) [inline service/csvService.go]: This logic results in doing work that you are not sure is needed.

I'd recommend to first check if record[0] == id, and if so, then append the corresponding doc.

```suggestion
		if record[0] == id {
			doc := Doc{
				Key:       record[0],
				Title:     record[1],
				Type:      record[2],
				Published: record[3],
			}
			
			recordsStruct = append(recordsStruct, doc)
		}

```

File: service/csvService.go [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	if len(records) == 0 {
+		fmt.Println("The file is empty")
```
Comment (RolloCasanova) [inline service/csvService.go]: This should be logged, instead of being printed.

File: container/inject.go [hunk 24]
```diff
@@ -0,0 +1,19 @@
+package container
+
+import (
+	"fmt"
+
+	"github.com/golobby/container"
+	"github.com/oscarSantoyo/academy-go-q12021/service"
+)
+
+func Connect ( ){
```
Comment (RolloCasanova) [inline container/inject.go]:
```suggestion
func Connect() {
```

File: container/inject.go [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+	fmt.Println("Connecting container")
+	// var instance = container.NewContainer();
+	container.Singleton(func () service.Search {
```
Comment (RolloCasanova) [inline container/inject.go]:
```suggestion
	container.Singleton(func() service.Search {
```

File: container/inject.go [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+		return &service.SearchImpl{}
+	})
+	container.Singleton(func () service.CsvService {
```
Comment (RolloCasanova) [inline container/inject.go]:
```suggestion
	container.Singleton(func() service.CsvService {
```

File: controller/search.go [hunk 27]
```diff
@@ -0,0 +1,26 @@
+package controller
+
+import (
+	"fmt"
+
+	"github.com/golobby/container"
+	"github.com/oscarSantoyo/academy-go-q12021/service"
+)
```
Comment (RolloCasanova) [inline controller/search.go]:
```suggestion
import (
	"fmt"
	
	"github.com/oscarSantoyo/academy-go-q12021/service"

	"github.com/golobby/container"
)
```

File: controller/search.go [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+
+var search service.Search
+
+func SearchById(id string) []service.Doc{
```
Comment (RolloCasanova) [inline controller/search.go]:
```suggestion
func SearchByID(id string) []service.Doc{
```

File: router/router.go [hunk 29]
```diff
@@ -0,0 +1,48 @@
+package router
+
+import (
+	"net/http"
+
+	"github.com/labstack/echo/v4"
+	"github.com/oscarSantoyo/academy-go-q12021/controller"
+)
```
Comment (RolloCasanova) [inline router/router.go]:
```suggestion
import (
	"net/http"
	
	"github.com/oscarSantoyo/academy-go-q12021/controller"

	"github.com/labstack/echo/v4"
)
```

File: main.go [hunk 30]
```diff
@@ -0,0 +1,26 @@
+package main
+
+import (
+	"github.com/labstack/echo/v4"
+	"github.com/labstack/echo/v4/middleware"
+	"github.com/oscarSantoyo/academy-go-q12021/container"
+	"github.com/oscarSantoyo/academy-go-q12021/router"
+)
```
Comment (RolloCasanova) [inline main.go]:
```suggestion
import (
	"github.com/oscarSantoyo/academy-go-q12021/container"
	"github.com/oscarSantoyo/academy-go-q12021/router"

	"github.com/labstack/echo/v4"
	"github.com/labstack/echo/v4/middleware"
)
```

**Review by RolloCasanova — COMMENTED** [review]
Please, add/apply suggested changes to your project.

Any doubt or inquiry don't hesitate to reach me out.

File: README.md (hunk 1 above)
Comment (oscarSantoyo) [reply README.md:35]: Added :+1:

File: container/inject.go (hunk 24 above)
Comment (oscarSantoyo) [reply container/inject.go]: :+1:

File: container/inject.go (hunk 3 above)
Comment (oscarSantoyo) [reply container/inject.go]: :+1:

File: container/inject.go (hunk 4 above)
Comment (oscarSantoyo) [reply container/inject.go]: :+1:

File: container/inject.go (hunk 25 above)
Comment (oscarSantoyo) [reply container/inject.go]: :+1:

File: container/inject.go (hunk 26 above)
Comment (oscarSantoyo) [reply container/inject.go]: :+1:

File: controller/search.go (hunk 27 above)
Comment (oscarSantoyo) [reply controller/search.go]: :+1:

File: controller/search.go (hunk 28 above)
Comment (oscarSantoyo) [reply controller/search.go]: :+1:

File: main.go (hunk 30 above)
Comment (oscarSantoyo) [reply main.go]: :+1:

File: main.go (hunk 5 above)
Comment (oscarSantoyo) [reply main.go]: :+1:

File: router/router.go (hunk 29 above)
Comment (oscarSantoyo) [reply router/router.go]: :+1:

File: service/csvService.go (hunk 15 above)
Comment (oscarSantoyo) [reply service/csvService.go]: :+1:

File: service/csvService.go (hunk 14 above)
Comment (oscarSantoyo) [reply service/csvService.go]: :+1:

File: service/csvService.go (hunk 12 above)
Comment (oscarSantoyo) [reply service/csvService.go]: :+1:

File: service/csvService.go (hunk 13 above)
Comment (oscarSantoyo) [reply service/csvService.go]: :+1:

File: service/csvService.go (hunk 20 above)
Comment (oscarSantoyo) [reply service/csvService.go]: :+1:

File: service/csvService.go (hunk 22 above)
Comment (oscarSantoyo) [reply service/csvService.go]: :+1: 
thanks!

File: service/csvService.go (hunk 23 above)
Comment (oscarSantoyo) [reply service/csvService.go]: :+1:

File: service/csvService.go (hunk 9 above)
Comment (oscarSantoyo) [reply service/csvService.go]: :+1:

File: service/csvService.go (hunk 18 above)
Comment (oscarSantoyo) [reply service/csvService.go]: :+1:

File: service/csvService.go (hunk 10 above)
Comment (oscarSantoyo) [reply service/csvService.go:96]: :+1:

File: service/csvService.go (hunk 19 above)
Comment (oscarSantoyo) [reply service/csvService.go:100]: :+1:

File: service/csvService.go (hunk 11 above)
Comment (oscarSantoyo) [reply service/csvService.go:109]: :+1:

File: service/csvService.go (hunk 16 above)
Comment (oscarSantoyo) [reply service/csvService.go]: :+1:

File: service/searchService.go (hunk 17 above)
Comment (oscarSantoyo) [reply service/searchService.go]: :+1:

**Pull request closed without merging**